rest/endpoint/v1/api.go

```diff
@@ -0,0 +1,893 @@
+	columnDefinitions := make([]m.ColumnDefinition, 0)
+	for _, col := range columns {
+		isStatic := false
+		if col.Kind == gocql.ColumnStatic {
+			isStatic = true
+		}
+
+		columnDefinitions = append(columnDefinitions, m.ColumnDefinition{
+			Name:           col.Name,
+			TypeDefinition: col.Type.Type().String(),
+			Static:         isStatic,
+		})
+	}
+	return columnDefinitions
+}
+
+func parseAndValidatePayload(obj interface{}, r *http.Request) error {
+	if err := json.NewDecoder(r.Body).Decode(obj); err != nil {
+		return err
+	}
+
+	if err := inputValidator.Struct(obj); err != nil {
+		return e.TranslateValidatorError(err, trans)
+	}
+
+	return nil
+}
+
+func (s *routeList) primaryKeyToString(m []types.ConditionItem) string {
+	jsonString, err := json.Marshal(m)
+	if err != nil {
+		s.logger.Debug("unable to convert primary key map to string", "error", err)
+	}
+
+	return string(jsonString)
+}
+
+// primaryKeyValues will return the partition columns of a table
+func primaryKeyValues(identifier string, tblMetadata *gocql.TableMetadata) ([]string, []interface{}, error) {
+	// Use array to maintain field order all the way
+	parts := strings.Split(identifier, ";")
+	columns := make([]string, 0, len(parts))
+	values := make([]interface{}, len(parts))
+	for i, key := range tblMetadata.PartitionKey {
+		if i >= len(parts) {
+			// In this case the table has a composite key but we were not passed in a value for each column
+			return nil, nil, errors.New("not enough parts provided for primary keys")
+		}
+		index := len(columns)
+		values[index] = parts[index]
+		columns = append(columns, key.Name)
+	}
+
+	// Add clustering keys when provided
+	for i := 0; i < len(tblMetadata.ClusteringColumns) && len(columns) < len(parts); i++ {
+		key := tblMetadata.ClusteringColumns[i]
+		index := len(columns)
+		values[index] = parts[index]
+		columns = append(columns, key.Name)
+	}
+
+	return columns, values, nil
+}
+
+func getCQLType(typeInfo gocql.TypeInfo, val interface{}) (interface{}, error) {
+	switch typeInfo.Type() {
+	// TODO: Share logic with graphql
+	case gocql.TypeDecimal:
+		d := new(inf.Dec)
+		if f, ok := val.(float64); ok {
+			d.SetString(fmt.Sprintf("%f", f))
+		} else if s, ok := val.(string); ok {
+			floatVal, err := strconv.ParseFloat(s, 64)
+			if err != nil {
+				return nil, errors.New("wrong value provided for decimal type")
+			}
+			d.SetString(fmt.Sprintf("%f", floatVal))
+		}
+		return d, nil
+	case gocql.TypeInt, gocql.TypeTinyInt, gocql.TypeSmallInt:
+		if f, ok := val.(float64); ok {
+			return int(f), nil
+		}
+
+		return nil, errors.New("wrong value provided for int type")
+	case gocql.TypeFloat:
+		if f, ok := val.(float64); ok {
+			return float32(f), nil
+		}
+
+		return nil, errors.New("wrong value provided for float type")
+	case gocql.TypeTimestamp:
+		if strValue, ok := val.(string); ok {
+			value := &time.Time{}
+			if err := value.UnmarshalText([]byte(strValue)); err != nil {
```

Comment:
Won't this fail on valid formats?

```go
// UnmarshalText implements the encoding.TextUnmarshaler interface.
// The time is expected to be in RFC 3339 format.
func (t *Time) UnmarshalText(data []byte) error {
```

Since the time package defines RFC-3339 as `RFC3339     = "2006-01-02T15:04:05Z07:00"` but C* accepts all of these

https://docs.datastax.com/en/cql-oss/3.x/cql/cql_reference/timestamp_type_r.html
```
yyyy-mm-dd HH:mm
yyyy-mm-dd HH:mm:ss
yyyy-mm-dd HH:mmZ
yyyy-mm-dd HH:mm:ssZ
yyyy-mm-dd'T'HH:mm
yyyy-mm-dd'T'HH:mmZ
yyyy-mm-dd'T'HH:mm:ss
yyyy-mm-dd'T'HH:mm:ssZ
yyyy-mm-dd'T'HH:mm:ss.ffffffZ
yyyy-mm-dd
yyyy-mm-ddZ
```

Reply:
The driver will serialize the `Time` structure into CQL timestamp representation (millis offset since unix epoch).

On the data apis we will be accepting only RFC-3339 formatted timestamp values, which is fine as long as we document it. We could add support for more formats in the future.

Reply:
Oh, got it. I was using the dateparse package so that I could be 1 to 1 with C* supported timestamps.